File: challenge_4/c/karanchawla/challenge_4.c
#include <stdio.h>
#include <stdlib.h> 

//Define node
typedef struct node
{
	/* data */
	int data;
	struct node* left, *right;
}Node;

//utility function for adding a new node
Node* newNode(int data)
{
	Node* newNode = (Node*) malloc(sizeof(Node));
	newNode->left = NULL;
	newNode->right = NULL;
	newNode->data = data;

	return newNode;
}

//function to invert the tree
void invertTree(Node* head)
{
	if(head==NULL)
	{
		return;
	}
	else
	{
		Node* temp = head;
		invertTree(head->left);
		invertTree(head->right);

		temp = head->left;
		head->left = head->right;
		head->right = temp;
	}

	return;
}

//utility function for inorder tree traversal
void inOrder(struct node* node) 
{
  if (node == NULL) 
    return;

  inOrder(node->left);
  printf("%d ", node->data);
  inOrder(node->right);
}  

//Driver function
int main(void)
{
	struct node *root = newNode(4);
	root->left        = newNode(2);
	root->right       = newNode(7);
	root->left->left  = newNode(6);
	root->left->right = newNode(9); 
	root->right->left  = newNode(1);
	root->right->right = newNode(3);

	inOrder(root);
	
	printf("\n");

	invertTree(root);

	inOrder(root);
	return 0;
}

File: challenge_4/python/system123/challenge4.py

class Node:

	def __init__(self, value):
		self.value = value
		self.left = None
		self.right = None


class BinTree:

	def __init__(self):
		self.root = None

	def add(self, value):
		node = Node(value)
		if self.root == None:
			self.root = node
		else:
			self._add(node, self.root)

	def get_root(self):
		return(self.root)

	def invert(self):
		self._swap(self.root)

	def _swap(self, parent):
		if parent == None:
			return

		tmp = parent.right
		parent.right = parent.left
		parent.left = tmp

		self._swap(parent.left)
		self._swap(parent.right)

	def _add(self, node, parent):
		if node.value == None:
			return
		elif node.value <= parent.value:
			if parent.left == None:
				parent.left = node
			else:
				self._add(node, parent.left)
		else:
			if parent.right == None:
				parent.right = node
			else:
				self._add(node, parent.right)

tree = BinTree()
tree.add(4)
tree.add(2)
tree.add(1)
tree.add(3)
tree.add(7)
tree.add(6)
tree.add(9)

print(tree.get_root().left.right.value)
tree.invert()
print(tree.get_root().left.right.value)


